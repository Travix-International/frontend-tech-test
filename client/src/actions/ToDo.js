import {
  INSERT_TODO,
  FETCH_ALL_TODO,
  INSERT_TODO_SUCCESS,
  FETCH_ALL_TODO_SUCCESS,
  SELECTED_TODO_SUCCESS,
  ON_TODO_UPDATE_SUCCESS,
  ON_BATCH_TODO_UPDATE,
  ON_BATCH_TODO_UPDATE_SUCCESS,
  GET_ALL_TODO,
  GET_NAV_FILTERS,
  GET_NAV_LABELS,
  GET_STARRED_TODO,
  GET_UNSELECTED_ALL_TODO,
  GET_UNSTARRED_TODO,
  HANDLE_REQUEST_CLOSE,
  ON_DELETE_TODO,
  ON_HIDE_LOADER,
  ON_LABEL_MENU_ITEM_SELECT,
  ON_LABEL_SELECT,
  ON_LABEL_UPDATE,
  ON_OPTION_MENU_SELECT,
  ON_SORTEND,
  ON_TODO_ADD,
  ON_TODO_CHECKED,
  ON_TODO_SELECT,
  ON_TODO_UPDATE,
  ON_TOGGLE_DRAWER,
  SEARCH_TODO,
  SELECT_ALL_TODO,
  SET_CURRENT_TODO_NULL,
  SHOW_MESSAGE,
  UPDATE_SEARCH,
  DELETE_TODO_SUCCESS
} from 'constants/ActionTypes'


export const fetchTodo = () => {
  return {
    type: FETCH_ALL_TODO
  };
};
export const insertTodo = (todo) => {
  return {
    type: INSERT_TODO,
    payload: todo
  };
};

export const fetchTodosSuccess = (mails) => {
  return {
    type: FETCH_ALL_TODO_SUCCESS,
    payload: mails
  }
};

export const insertTodoSuccess = (data) => {
  return {
    type: INSERT_TODO_SUCCESS,
    payload: data
  };
};

export const selectTodoSuccess = (data) => {
  return {
    type: SELECTED_TODO_SUCCESS,
    payload: data
  };
};

export const updateTodoSuccess = (data) => {
  return {
    type: ON_TODO_UPDATE_SUCCESS,
    payload: data
  };
};

export const updateBatchTodoSuccess = (data) => {
  return {
    type: ON_BATCH_TODO_UPDATE_SUCCESS,
    payload: data
  };
};


export const onTodoBatchUpdate = (data) => {
  return {
    type: ON_BATCH_TODO_UPDATE,
    payload: data
  };
};

export const deleteTodoSuccess = (data) => {
  return {
    type: DELETE_TODO_SUCCESS,
    payload: data
  };
};

export const showTodoMessage = (message) => {
  return {
    type: SHOW_MESSAGE,
    payload: message
  };
};
export const onSortEnd = (data) => {
  return {
    type: ON_SORTEND,
    payload: data
  };
};

export const onOptionMenuSelect = () => {
  return {
    type: ON_OPTION_MENU_SELECT,
  };
};
export const onLabelSelect = () => {
  return {
    type: ON_LABEL_SELECT,
  };
};
export const selectAllTodo = () => {
  return {
    type: SELECT_ALL_TODO,
  };
};
export const getAllTodo = () => {
  return {
    type: GET_ALL_TODO,
  };
};


export const getUnselectedAllTodo = () => {
  return {
    type: GET_UNSELECTED_ALL_TODO,
  };
};

export const getStarredToDo = () => {
  return {
    type: GET_STARRED_TODO,
  };
};

export const getUnStarredTodo = () => {
  return {
    type: GET_UNSTARRED_TODO,
  };
};

export const onLabelMenuItemSelect = (label) => {
  return {
    type: ON_LABEL_MENU_ITEM_SELECT,
    payload: label
  };
};
export const onLabelUpdate = (data) => {
  return {
    type: ON_LABEL_UPDATE,
    payload: data
  };
};

export const onToDoUpdate = (data) => {
  return {
    type: ON_TODO_UPDATE,
    payload: data
  };
};

export const onDeleteToDo = (data) => {
  return {
    type: ON_DELETE_TODO,
    payload: data
  };
};
export const getNavFilters = (data) => {
  return {
    type: GET_NAV_FILTERS,
    payload: data
  };
};
export const getNavLabels = (data) => {
  return {
    type: GET_NAV_LABELS,
    payload: data
  };
};

export const onSearchTodo = (searchText) => {
  return {
    type: SEARCH_TODO,
    payload: searchText
  };
};
export const onTodoChecked = (data) => {
  return {
    type: ON_TODO_CHECKED,
    payload: data
  };
};
export const onTodoAdd = (data) => {
  return {
    type: ON_TODO_ADD,
    payload: data
  };
};
export const onTodoSelect = (todo) => {
  return {
    type: ON_TODO_SELECT,
    payload: todo
  };
};
export const setCurrentToDoNull = () => {
  return {
    type: SET_CURRENT_TODO_NULL,
  };
};

export const toDoToggleDrawer = () => {
  return {
    type: ON_TOGGLE_DRAWER,
  };
};

export const updateSearch = (searchTodo) => {
  return {
    type: UPDATE_SEARCH,
    payload: searchTodo
  };
};

export const hideToDoLoader = () => {
  return {
    type: ON_HIDE_LOADER,
  };
};
export const handleToDoMenuRequestClose = () => {
  return {
    type: HANDLE_REQUEST_CLOSE,
  };
};

