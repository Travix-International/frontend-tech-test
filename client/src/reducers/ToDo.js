import arrayMove from 'array-move';
import {
  FETCH_ALL_TODO_SUCCESS,
  INSERT_TODO_SUCCESS,
  DELETE_TODO_SUCCESS,
  ON_BATCH_TODO_UPDATE_SUCCESS,
  GET_ALL_TODO,
  GET_NAV_FILTERS,
  GET_NAV_LABELS,
  GET_STARRED_TODO,
  GET_UNSELECTED_ALL_TODO,
  GET_UNSTARRED_TODO,
  HANDLE_REQUEST_CLOSE,
  ON_HIDE_LOADER,
  ON_LABEL_MENU_ITEM_SELECT,
  ON_LABEL_SELECT,
  ON_LABEL_UPDATE,
  ON_OPTION_MENU_ITEM_SELECT,
  ON_OPTION_MENU_SELECT,
  ON_SORTEND,
  ON_TODO_ADD,
  ON_TODO_CHECKED,
  SELECTED_TODO_SUCCESS,
  ON_TOGGLE_DRAWER,
  SEARCH_TODO,
  SELECT_ALL_TODO,
  SET_CURRENT_TODO_NULL,
  SHOW_MESSAGE,
  SHOW_TODOS,
  UPDATE_SEARCH,
  ON_TODO_UPDATE_SUCCESS
} from 'constants/ActionTypes'

const INIT_STATE = {
  searchTodo: '',
  alertMessage: '',
  loader: true,
  showMessage: false,
  optionName: 'None',
  drawerState: false,
  allToDos: [],
  currentTodo: null,
  selectedToDos: 0,
  labelMenuState: false,
  optionMenuState: false,
  toDos: [],
  filter: -1
};


export default (state = INIT_STATE, action) => {
  switch (action.type) {
    case ON_SORTEND: {
      return {
        ...state, toDos: arrayMove(state.toDos, action.payload.oldIndex, action.payload.newIndex)
      }
    }
    case ON_LABEL_SELECT: {
      return {
        ...state,
        labelMenuState: !state.labelMenuState
      }
    }
    case ON_OPTION_MENU_SELECT: {
      return {
        ...state,
        optionMenuState: !state.optionMenuState
      }
    }
    case ON_OPTION_MENU_ITEM_SELECT: {
      return {...state}
    }
    case SELECT_ALL_TODO: {
      let toDos = state.allToDos.map((todo) => todo ? {
        ...todo,
        selected: true
      } : todo);

      return {
        ...state,
        selectedToDos: toDos.length,
        allToDos: toDos,
        optionName: 'All',
        toDos: toDos
      }
    }
    case GET_ALL_TODO: {
      return {
        ...state,
        currentTodo: null,
        toDos: state.allToDos
      }
    }
    case GET_UNSELECTED_ALL_TODO: {

      let toDos = state.allToDos.map((todo) => todo ? {
        ...todo,
        selected: false
      } : todo);
      return {
        ...state,
        selectedToDos: 0,
        allToDos: toDos,
        optionName: 'None',
        toDos: toDos
      }
    }
    case GET_STARRED_TODO: {
      let selectedToDos = 0;
      let toDos = state.allToDos.map((todo) => {
        if (todo.starred) {
          selectedToDos++;
          return {...todo, selected: true};
        }
        return {...todo, selected: false}
      });
      return {
        ...state,
        selectedToDos: selectedToDos,
        allToDos: toDos,
        optionName: 'Stared',
        toDos: toDos.filter(todo => !todo.deleted)
      }
    }
    case GET_UNSTARRED_TODO: {
      let selectedToDos = 0;
      let toDos = state.allToDos.map((todo) => {
        if (!todo.starred) {
          selectedToDos++;
          return {...todo, selected: true};
        }
        return {...todo, selected: false}
      });
      return {
        ...state,
        selectedToDos: selectedToDos,
        allToDos: toDos,
        optionName: 'UnStarred',
        toDos: toDos.filter(todo => !todo.deleted)
      }
    }
  

    case ON_LABEL_MENU_ITEM_SELECT: {
      const toDos = state.allToDos.map(todo => {
          if (todo.selected) {
            if (todo.labels.includes(action.payload.id)) {
              todo.labels.splice(todo.labels.indexOf(action.payload.id), 1);
              return {...todo, labels: todo.labels};
            } else {
              return {...todo, labels: todo.labels.concat(action.payload.id)};
            }
          } else {
            return todo;
          }
        }
      );
      return {
        ...state,
        alertMessage: 'Label Updated Successfully',
        showMessage: true,
        allToDos: toDos,
        toDos: toDos
      }
    }
    case ON_LABEL_UPDATE: {
      if (action.payload.data.labels.includes(action.payload.label.id)) {
        action.payload.data.labels.splice(action.payload.data.labels.indexOf(action.payload.label.id), 1);
      } else {
        action.payload.data.labels = action.payload.data.labels.concat(action.payload.label.id);
      }
      const toDos = state.allToDos.map(todo => {
          if (todo.id === action.payload.data.id) {
            return action.payload.data;
          } else {
            return todo;
          }
        }
      );

      return {
        ...state,
        alertMessage: 'Label Updated Successfully',
        showMessage: true,
        currentTodo: action.payload,
        allToDos: toDos,
        toDos: toDos,
      }
    }

    case ON_TODO_UPDATE_SUCCESS: {
      const toDos = state.allToDos.map(todo => {
        if (todo.id === action.payload.data.id) {
          return action.payload.data;
        } else {
          return todo;
        }
      });
      return {
        ...state,
        alertMessage: action.payload.message,
        showMessage: true,
        currentTodo: action.payload,
        allToDos: toDos,
        toDos: toDos,
      }
    }
    case ON_BATCH_TODO_UPDATE_SUCCESS: {
      const toDos = state.allToDos.map(todo => {
        for(var i = 0; i < action.payload.data.length; i++) {
          if (todo.id === action.payload.data[i].id) {
            return action.payload.data[i];
          }
      }
      return todo;
    });
      return {
        ...state,
        alertMessage: "Updated Successfully",
        showMessage: true,
        currentTodo: state.selectedToDos.length > 0 || state.currentTodo == null ? null : action.payload.data[0],
        allToDos: toDos,
        toDos: toDos,
      }
    }
    case DELETE_TODO_SUCCESS: {
      let selectedToDos = 0;
      const toDos = state.allToDos.map(todo => {
          if (todo.selected) {
            selectedToDos++
          }
          if (action.payload.id === todo.id) {
            if (todo.selected) {
              selectedToDos--
            }
            return {...todo, deleted: true};
          } else {
            return todo;
          }
        }
      );
      
      return {
        ...state,
        alertMessage: action.payload.message,
        showMessage: true,
        allToDos: toDos.filter((todo) => !todo.deleted),
        currentTodo: null,
        selectedToDos: selectedToDos,
        toDos: toDos.filter((todo) => !todo.deleted),
        loader: false,
      }
    }

    case GET_NAV_LABELS: {
      const filterMails = state.allToDos.filter(todo => todo.labels.includes(action.payload.id));
      return {
        ...state,
        currentTodo: null,
        toDos: filterMails
      }
    }
    case GET_NAV_FILTERS: {
      const filterTodos = state.allToDos.filter(todo => 
        (action.payload.id === 0 && todo.starred) ||
        (action.payload.id === 1 && !todo.completed) ||
        (action.payload.id === 2 && todo.completed))
      return {
        ...state,
        currentTodo: null,
        filter: action.payload.id,
        toDos: filterTodos
      }
    }

    case SEARCH_TODO: {
      if (action.payload === '') {
        return {...state, toDos: state.allToDos.filter((todo) => !todo.deleted)};
      } else {
        const searchToDos = state.allToDos.filter((todo) =>
          !todo.deleted && todo.title.toLowerCase().indexOf(action.payload.toLowerCase()) > -1);
        return {...state, toDos: searchToDos}
      }
    }
    case SHOW_TODOS: {
      return {...state}
    }
    case ON_TODO_CHECKED: {
      action.payload.selected = !action.payload.selected;
      let selectedToDos = 0;
      const toDos = state.toDos.map(todo => {
          if (todo.selected) {
            selectedToDos++;
          }
          if (todo.id === action.payload.id) {
            if (todo.selected) {
              selectedToDos++;
            }
            return action.payload;
          } else {
            return todo;
          }
        }
      );
      return {
        ...state,
        selectedToDos: selectedToDos,
        toDos: toDos
      }
    }
    case ON_TODO_ADD: {
      return {
        ...state,
        alertMessage: 'ToDo Added Successfully',
        showMessage: true,
        toDos: state.allToDos.concat(action.payload),
        allToDos: state.allToDos.concat(action.payload)
      }
    }
    case SELECTED_TODO_SUCCESS: {
    
      return {
        ...state,
        currentTodo: action.payload.task,
        loader: false
      }
    }
    case SET_CURRENT_TODO_NULL: {
      return {...state, currentTodo: null}
    }
    case UPDATE_SEARCH: {
      return {...state, searchTodo: action.payload}
    }
    case ON_TOGGLE_DRAWER: {
      return {...state, drawerState: !state.drawerState}
    }
    case HANDLE_REQUEST_CLOSE: {
      return {...state, showMessage: false, addTodo: false, labelMenuState: false, optionMenuState: false}
    }
    case ON_HIDE_LOADER: {
      return {...state, loader: false}
    }
    case FETCH_ALL_TODO_SUCCESS: {
      return {
        ...state,
        allToDos: action.payload,
        toDos: action.payload,
        loader: false,
      }
    }
    case INSERT_TODO_SUCCESS: {
      return {
        ...state,
        alertMessage: action.payload.message,
        showMessage: true,
        allToDos: state.allToDos.concat(action.payload.data),
        toDos: state.allToDos.concat(action.payload.data),
        loader: false,
      }
    }
    
    case SHOW_MESSAGE: {
      return {
        ...state,
        alertMessage: action.payload,
        showMessage: true,
        loader: false
      }
    }

    default:
      return state;
  }
}