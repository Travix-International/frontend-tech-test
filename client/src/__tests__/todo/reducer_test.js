import ToDo, { INIT_STATE } from '../../reducers/ToDo';

describe('ToDo Reducer', () => {

  it('Should return the initial state', () => {
    expect(ToDo(undefined, {})).toMatchSnapshot();
  });

  it('Should handle ON_TODO_ADD', () => {
    expect(
      ToDo(INIT_STATE,
        {
          type: 'ON_TODO_ADD',
        }),
    ).toMatchSnapshot();
  });

  it('Should handle GET_ALL_TODO', () => {
    expect(
      ToDo(INIT_STATE,
        {
          type: 'GET_ALL_TODO',
        }),
    ).toMatchSnapshot();
  });

  it('Should handle ON_TODO_UPDATE_SUCCESS', () => {
    expect(
      ToDo(INIT_STATE,
        {
          type: 'ON_TODO_UPDATE_SUCCESS',
        }),
    ).toMatchSnapshot();
  });

  it('Should handle DELETE_TODO_SUCCESS', () => {
    expect(
      ToDo(INIT_STATE,
        {
          type: 'DELETE_TODO_SUCCESS',
        }),
    ).toMatchSnapshot();
  });

  it('Should handle FETCH_ALL_TODO_SUCCESS', () => {
    expect(
      ToDo(INIT_STATE,
        {
          type: 'FETCH_ALL_TODO_SUCCESS',
        }),
    ).toMatchSnapshot();
  });

  it('Should handle INSERT_TODO_SUCCESS', () => {
    expect(
      ToDo(INIT_STATE,
        {
          type: 'INSERT_TODO_SUCCESS',
        }),
    ).toMatchSnapshot();
  });

});
