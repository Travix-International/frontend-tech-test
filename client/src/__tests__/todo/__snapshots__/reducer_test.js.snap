// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo Reducer Should handle DELETE_TODO_SUCCESS 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;

exports[`ToDo Reducer Should handle FETCH_ALL_TODO_SUCCESS 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;

exports[`ToDo Reducer Should handle GET_ALL_TODO 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;

exports[`ToDo Reducer Should handle INSERT_TODO_SUCCESS 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;

exports[`ToDo Reducer Should handle ON_TODO_ADD 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;

exports[`ToDo Reducer Should handle ON_TODO_UPDATE_SUCCESS 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;

exports[`ToDo Reducer Should return the initial state 1`] = `
Object {
  "alertMessage": "",
  "allToDos": Array [],
  "currentTodo": null,
  "drawerState": false,
  "filter": -1,
  "labelMenuState": false,
  "loader": true,
  "optionMenuState": false,
  "optionName": "None",
  "searchTodo": "",
  "selectedToDos": 0,
  "showMessage": false,
  "toDos": Array [],
}
`;
